### PR TITLE
Add redeploy to workload API and integration test

### PR DESCRIPTION
**Pull request by shanewxy**
Exposed redeploy REST API to the workload which set sets `cattle.io/timestamp` annotation to the current time.

Related issue: https://github.com/rancher/rancher/issues/23053
Related types PR: https://github.com/rancher/types/pull/1066

### CODE REVIEW

File: pkg/api/customization/workload/workload.go [hunk 1]
```diff
@@ -64,6 +67,9 @@ func (a ActionWrapper) ActionHandler(actionName string, action *types.Action, ap
 			return httperror.NewAPIError(httperror.InvalidAction, fmt.Sprintf("Pause deployment %s before resume", deployment.ID))
 		}
 		return updatePause(apiContext, false, deployment, "resume")
+	case "redeploy":
+		timestamp := time.Now().UTC().Format(protocol.ISO8601TimeFormat)
```
Comment (gitlawr) [inline pkg/api/customization/workload/workload.go]: Can we use a format in the standard library(https://golang.org/pkg/time/)
As the package name suggested `github.com/aws/aws-sdk-go/private/protocol` is meant to be used internally so we should not depend on it.

File: pkg/api/customization/workload/workload.go [hunk 2]
```diff
@@ -236,6 +252,7 @@ func DeploymentFormatter(apiContext *types.APIContext, resource *types.RawResour
 	resource.Actions["pause"] = apiContext.URLBuilder.ActionLinkByID(workloadSchema, workloadID, "pause")
 	resource.Actions["resume"] = apiContext.URLBuilder.ActionLinkByID(workloadSchema, workloadID, "resume")
 	resource.Actions["rollback"] = apiContext.URLBuilder.ActionLinkByID(workloadSchema, workloadID, "rollback")
+	resource.Actions["redeploy"] = apiContext.URLBuilder.ActionLinkByID(workloadSchema, workloadID, "redeploy")
```
Comment (gitlawr) [inline pkg/api/customization/workload/workload.go]: This only applies to deployments. We can also redeploy a statefulset/daemonset/replicaset in the UI. Can you check the complete list and make it work for all those types?

File: tests/integration/suite/test_workloads.py [hunk 3]
```diff
@@ -310,6 +311,34 @@ def test_statefulset_workload_volumemount_subpath(admin_pc):
...
+        annotations={
+            "cattle.io/timestamp":
+                datetime.datetime.now().isoformat()}, )
+    p_client.action(workload, "redeploy")
+    timestamp = p_client.list_workload(
+        uuid=workload.uuid).data[0].annotations["cattle.io/timestamp"]
+    assert timestamp != workload.annotations["cattle.io/timestamp"]
+    p_client.delete(workload)
+    admin_pc.cluster.client.delete(ns)
+    admin_mc.client.delete(project)
```
Comment (gitlawr) [inline tests/integration/suite/test_workloads.py]: The admin_pc fixture handles the project creation/deletion so we don't need to do that in the tests

File: pkg/api/customization/workload/workload.go [hunk 4]
```diff
@@ -203,7 +206,19 @@ func (a ActionWrapper) rollbackDeployment(apiContext *types.APIContext, clusterC
 	}
 	return nil
 }
-
+func updateTimestamp(apiContext *types.APIContext, workload projectclient.Workload) error {
+	timestamp := time.Now().UTC().Format(time.RFC3339)
+	data, err := convert.EncodeToMap(workload)
+	if err != nil {
+		return httperror.WrapAPIError(err, httperror.ServerError, "Failed to parse workload")
```
Comment (mrajashree) [inline pkg/api/customization/workload/workload.go]: error message should be "Failed to redeploy workload"?

File: pkg/api/customization/workload/workload.go [hunk 5]
```diff
@@ -203,7 +206,19 @@ func (a ActionWrapper) rollbackDeployment(apiContext *types.APIContext, clusterC
...
+	}
+	values.PutValue(data, timestamp, "annotations", "cattle.io/timestamp")
+	err = update(apiContext, data, workload.ID)
+	if err != nil {
+		return httperror.NewAPIError(httperror.ServerError, fmt.Sprintf("Error updating workload %s by redeploy : %s", workload.ID, err.Error()))
```
Comment (mrajashree) [inline pkg/api/customization/workload/workload.go]: can we say "Error redeploying workload" instead?

File: tests/integration/suite/test_workloads.py [hunk 6]
```diff
@@ -310,6 +310,31 @@ def test_statefulset_workload_volumemount_subpath(admin_pc):
         assert e.value.error.status == 422
 
 
+def test_workload_redeploy(admin_pc):
+    client = admin_pc.client
+    ns = admin_pc.cluster.client.create_namespace(
+        name=random_str(),
+        projectId=admin_pc.project.id)
```
Comment (mrajashree) [inline tests/integration/suite/test_workloads.py]: can we add `remove_resource` for this namespace and the workload on line 319 too, we won't have to use client.delete

File: tests/integration/suite/test_workloads.py [hunk 7]
```diff
@@ -310,6 +310,31 @@ def test_statefulset_workload_volumemount_subpath(admin_pc):
...
+        scale=1,
+        containers=[{
+            'name': 'one',
+            'image': 'nginx',
+        }],
+        annotations={
+            "cattle.io/timestamp":
+                datetime.datetime.now().isoformat()}, )
+    client.action(workload, "redeploy")
+    timestamp = client.list_workload(
```
Comment (mrajashree) [inline tests/integration/suite/test_workloads.py]: do we need some retry/sleep/check logic around this? might lead to error in drone

File: pkg/api/customization/workload/workload.go [hunk 8]
```diff
@@ -224,6 +239,11 @@ func Formatter(apiContext *types.APIContext, resource *types.RawResource) {
 	resource.Links["self"] = apiContext.URLBuilder.ResourceLinkByID(workloadSchema, workloadID)
 	resource.Links["remove"] = apiContext.URLBuilder.ResourceLinkByID(workloadSchema, workloadID)
 	resource.Links["update"] = apiContext.URLBuilder.ResourceLinkByID(workloadSchema, workloadID)
+	resourceType := resource.Type
+	//add redeploy action to the workload types that support redeploy
```
Comment (mrajashree) [inline pkg/api/customization/workload/workload.go]: maybe nit: can we have a map of workloads that allow redeploy instead and check with that?

**Comment by shanewxy** [comment]
@mrajashree updated

**Comment by shanewxy** [comment]
Updated to latest rancher types

File: tests/integration/suite/test_workloads.py [hunk 9]
```diff
@@ -341,3 +369,19 @@ def wait_for_service_cluserip_reset(client, name, timeout=30):
         if time.time() - start > timeout:
             raise Exception('Timeout waiting for workload service')
     return services.data[0]
+
+
+def wait_for_workload_active(client, workload, timeout=30):
```
Comment (dramich) [inline tests/integration/suite/test_workloads.py]: This is a duplication of code that already lives in `wait_for`, use that and pass in your callback and fail handler.

File: tests/integration/suite/test_workloads.py [hunk 10]
```diff
@@ -310,6 +310,34 @@ def test_statefulset_workload_volumemount_subpath(admin_pc):
...
+            "cattle.io/timestamp":
+                datetime.datetime.now().isoformat()}, )
+    remove_resource(workload)
+    workload = wait_for_workload_active(client, workload)
+    client.action(workload, "redeploy")
+    timestamp = client.list_workload(
+        uuid=workload.uuid).data[0].annotations["cattle.io/timestamp"]
+    assert timestamp != workload.annotations["cattle.io/timestamp"]
+    client.delete(workload)
+    admin_pc.cluster.client.delete(ns)
```
Comment (dramich) [inline tests/integration/suite/test_workloads.py]: Both of these delete calls are unnecessary, `remove_resource` takes care of this and will handle errors.

**Comment by shanewxy** [comment]
@dramich updated

File: tests/integration/suite/test_workloads.py (hunk 7 above)
Comment (rmweir) [reply tests/integration/suite/test_workloads.py]: I agree with @mrajashree this really needs retry as the annotation may not be set right away

File: tests/integration/suite/test_workloads.py [hunk 11]
```diff
@@ -310,6 +312,39 @@ def test_statefulset_workload_volumemount_subpath(admin_pc):
...
+        annotations={
+            "cattle.io/timestamp":
+                datetime.datetime.now().isoformat()}, )
+    remove_resource(workload)
+
+    def workload_active():
+        workloads = client.list_workload(uuid=workload.uuid).data
+        return workloads[0].state == 'active'
+
+    wait_for(workload_active,
```
Comment (rmweir) [inline tests/integration/suite/test_workloads.py]: I think we can already observe that the default time in wait_for (45 seconds) is not long enough. See these two breaks: https://drone-pr.rancher.io/rancher/rancher/6875, https://drone-publish.rancher.io/rancher/rancher/2202

**Pull request merged**